components/browser/storage-sync/src/main/java/mozilla/components/browser/storage/sync/PlacesHistoryStorage.kt

```diff
@@ -165,11 +171,69 @@ open class PlacesHistoryStorage(context: Context) : PlacesStorage(context), Hist
     override suspend fun sync(authInfo: AuthInfo): SyncStatus {
         return try {
             withContext(scope.coroutineContext) {
-                places.syncHistory(authInfo.into())
+                val ping = places.syncHistory(authInfo.into())
+                submitHistoryPing(ping)
                 SyncStatus.Ok
             }
         } catch (e: PlacesException) {
             SyncStatus.Error(e)
         }
     }
 }
+
+@Suppress("ComplexMethod", "NestedBlockDepth")
+private fun submitHistoryPing(ping: SyncTelemetryPing) {
+    ping.syncs.forEach eachSync@{ sync ->
+        sync.failureReason?.let {
+            recordFailureReason(it)
+            return@eachSync
+        }
+        sync.engines.forEach eachEngine@{ engine ->
+            if (engine.name != "history") {
+                return@eachEngine
+            }
+            HistorySync.apply {
+                val base = BaseGleanSyncPing.fromEngineInfo(ping.uid, engine)
+                uid.set(base.uid)
+                startedAt.set(base.startedAt)
+                finishedAt.set(base.finishedAt)
+                if (base.applied > 0) {
+                    // Since all Sync ping counters have `lifetime: ping`, and
+                    // we send the ping immediately after, we don't need to
+                    // reset the counters before calling `add`.
+                    incoming["applied"].add(base.applied)
+                }
+                if (base.failedToApply > 0) {
+                    incoming["failed_to_apply"].add(base.failedToApply)
+                }
+                if (base.reconciled > 0) {
+                    incoming["reconciled"].add(base.reconciled)
+                }
+                if (base.uploaded > 0) {
+                    outgoing["uploaded"].add(base.uploaded)
+                }
+                if (base.failedToUpload > 0) {
+                    outgoing["failed_to_upload"].add(base.failedToUpload)
+                }
+                if (base.outgoingBatches > 0) {
+                    outgoingBatches.add(base.outgoingBatches)
+                }
+                base.failureReason?.let {
+                    recordFailureReason(it)
+                }
+            }
+            Pings.historySync.send()
```

Comment:
After reading the discussion in Slack, I realized that we shouldn't need to trigger the ping at all. Coming back to this code, I think there's an alternative solution: what if refactored `Pings.historySync.send()` to a separate, `internal` function, something like this:

```
internal fun submitHystorySyncPing() {
  Pings.historySync.send()
}
```

Then in tests, since you're testing each engine in isolation, you could probably do something like:

```
// TODO: mock submitHystorySyncPing

// Ensure that the ping submission code was called. 
verify(storage).submitHystorySyncPing()

// TODO: use the testing api
```

Alternatively, you could also change `MockingPlacesHistoryStorage` to accept a callback function and call that instead of `submitHystorySyncPing`.

Reply:
Thanks, I went with a slightly modified variation of your suggestion! 😄

Reply:
Great! Glad it worked out!